olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp

```diff
@@ -517,6 +520,148 @@ bool VersionedLayerClientImpl::RemoveFromCache(const geo::TileKey& tile) {
+                           AggregatedDataResponseCallback callback) {
+    auto catalog = catalog_;
+    auto layer_id = layer_id_;
+    auto settings = settings_;
+    auto pending_requests = pending_requests_;
+
+    auto data_task =
+        [=](client::CancellationContext context) -> AggregatedDataResponse {
+      auto version_response = GetVersion(request.GetBillingTag(),
+                                         request.GetFetchOption(), context);
+      if (!version_response.IsSuccessful()) {
+        return version_response.GetError();
+      }
+
+      auto version = version_response.GetResult().GetVersion();
+
+      // 1. Download quad tree index depth 4
+      // TODO: load from cache
+      auto fetch_option = request.GetFetchOption();
+      const auto& tile_key = request.GetTileKey();
+      const auto& root_tile_key =
+          tile_key.ChangedLevelBy(-kAggregateQuadTreeDepth);
+      auto root_tile = root_tile_key.ToHereTile();
+
+      auto query_api = ApiClientLookup::LookupApi(
+          catalog, context, "query", "v1", request.GetFetchOption(), settings);
+
+      if (!query_api.IsSuccessful()) {
+        OLP_SDK_LOG_ERROR(kLogTag,
+                          "QueryPartitionForVersionedTile: LookupApi failed.");
+        return query_api.GetError();
+      }
+
+      auto quad_tree_response = QueryApi::QuadTreeIndex(
+          query_api.GetResult(), layer_id, version, root_tile,
+          kAggregateQuadTreeDepth, boost::none, request.GetBillingTag(),
+          context);
+
+      if (!quad_tree_response.IsSuccessful()) {
+        OLP_SDK_LOG_ERROR_F(
+            kLogTag, "QuadTreeIndex failed (%s, %" PRId64 ", %" PRId32 ")",
+            root_tile.c_str(), version, kAggregateQuadTreeDepth);
+        return quad_tree_response.GetError();
+      }
+
+      const auto& quad_tree = quad_tree_response.GetResult();
+      const auto& subquads = quad_tree.GetSubQuads();
+
+      if (subquads.empty()) {
+        OLP_SDK_LOG_ERROR_F(
+            kLogTag, "QuadTreeIndex is empty (%s, %" PRId64 ", %" PRId32 ")",
+            root_tile.c_str(), version, kAggregateQuadTreeDepth);
+        return {{client::ErrorCode::NotFound, "Quad tree is empty."}};
+      }
+
+      // 2. Now we can look for requested tile in the quad tree or its closest
+      // ancestor
+      auto fetch_tile_key = geo::TileKey();
+      std::shared_ptr<model::SubQuad> fetch_subquad = nullptr;
+      for (const auto& sub : subquads) {
+        const auto key =
+            root_tile_key.AddedSubkey64(std::stoul(sub->GetSubQuadKey()));
+
+        const auto is_requested_tile = (key == tile_key);
+        const auto is_nearest_parent =
+            tile_key.IsChildOf(key) && key.Level() <= fetch_tile_key.Level();
+
+        if (is_nearest_parent || is_requested_tile) {
+          fetch_tile_key = key;
+          fetch_subquad = sub;
+        }
+
+        if (is_requested_tile) {
+          break;
+        }
+      }
+
```

Comment:
TODO: in case there is no nearest parent in subquads we need to search for nearest in parentQuads